File: python/pythonhpc/code/euler39/ctypes/euler39.c
#include <stdio.h>

int find_num_solutions(int p){
    int a,b,c,n;
    n=0;
    for(a=1;a<p/2;a++){
        for (b=a;b<p/2;b++){
            c=p-a-b;
            if(a*a+b*b==c*c){
                n=n+1;
            }
        }
    }
return n;
}

int main(){

int i;
int nrange;
int nsols;
int nmax,imax;
//nrange=1000;
nrange=5000;
nmax=0;
imax=0;


for (i=0;i<nrange;i++){
    nsols=find_num_solutions(i);
    if(nsols>nmax){
        nmax=nsols;
        imax=i;
    }
}

printf("imax=%d nmax=%d \n",imax,nmax);
return 0;
}

File: python/pythonhpc/code/euler39/cython/euler39.py
import time
import find_num_solutions

assert find_num_solutions.find_num_solutions(10)==0,"should have zero solutions for p=10"
assert find_num_solutions.find_num_solutions(12)==1,"should have one solution for p=12, as 3*3+4*4=5*5 and 3+4+5=12"

nrange=1000
# nrange = 5000
nmax=0
imax=0

start= time.time()
for i in range(1,nrange):
    nsols=find_num_solutions.find_num_solutions(i)
    if(nsols>nmax):
        nmax=nsols
        imax=i
end = time.time()

print "loop up to ", nrange , "took ", end - start, "seconds"

assert imax==840, "not getting the right answer"
assert nmax==8, "not getting the right number of solutions"
print imax,nmax
